### PR TITLE
Httpserver metric

### PR DESCRIPTION
**Pull request by interxd**

### CODE REVIEW

File: sentinel-core/log/metric/metric_node.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <stdint.h>
+#include <string>
+
+namespace Sentinel {
+namespace Metric {
+
+class MetricNode {
```
Comment (sczyh30) [inline sentinel-core/log/metric/metric_node.h]: @zyfjeff I've implemented the `MetricNode` class as `MetricItem` in master branch. After merging this PR we can replace the legacy `MetricNode` with `MetricItem` class.

File: sentinel-core/log/metric_writer.cc [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+  auto tp = std::chrono::system_clock::from_time_t(std::mktime(&tm));
+  time_second_base_ =
+      std::chrono::duration_cast<std::chrono::seconds>(tp.time_since_epoch())
+          .count();
+
+  pid_ = ::getpid();
+}
+
+/**
+ * 如果传入了time，就认为nodes中所有的时间时间戳都是time.
```
Comment (zyfjeff) [inline sentinel-core/log/metric_writer.cc]: Please use the english comment

File: sentinel-core/log/metric_writer.cc [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  // first write, should create file
+  if (!metric_out_.is_open()) {
+    base_file_name_ = FormMetricFileName(app_name, pid_);
+    CloseAndNewFile(NextFileNameOfDay(time));
+  }
+
+  auto second = time / 1000;
+
+  if (second < last_second_) {
+    // 时间靠前的直接忽略，不应该发生。
```
Comment (zyfjeff) [inline sentinel-core/log/metric_writer.cc]: ditto

File: sentinel-core/log/metric/metric_node.h (hunk 1 above)
Comment (zyfjeff) [reply sentinel-core/log/metric/metric_node.h]: Is it better to merge the master directly in this PR and modify it?

Comment (sczyh30) [reply sentinel-core/log/metric/metric_node.h]: Yes, it's more recommended to modify this in this PR.

File: sentinel-core/log/record_log.h [hunk 4]
```diff
@@ -2,14 +2,34 @@
...
+
 class RecordLog {
  public:
   RecordLog() = delete;
 
-  static void Info(std::string& message);
+  // TODO you must config record_log path when app startup
+  static void Init(const std::string& file_path);
+
+  static void Info(const std::string& message);
```
Comment (zyfjeff) [inline sentinel-core/log/record_log.h]: Is it better to use macros to replace these static methods?

File: sentinel-core/transport/command/BUILD [hunk 5]
```diff
@@ -0,0 +1,21 @@
+load("//bazel:copts.bzl", "DEFAULT_COPTS", "TEST_COPTS")
+
+cc_library(
+    name = "transport_httpserver",
```
Comment (zyfjeff) [inline sentinel-core/transport/command/BUILD]: Please split this library, this library is too big.the purpose of doing this is to make our dependence clearer.

File: sentinel-core/transport/command/command_request.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  CommandRequest& AddParam(const std::string& key, const std::string& value);
+  std::string GetParam(const std::string& key);
+  std::string GetParam(const std::string& key, const std::string& defaultValue);
+  const std::map<std::string, std::string>& GetParameters() const;
+
+  CommandRequest& AddMetadata(const std::string& key, const std::string& value);
+  std::string GetMetadata(const std::string& key) const;
+
+ public:
+  static const std::string kRequestTarget;
```
Comment (zyfjeff) [inline sentinel-core/transport/command/command_request.h]: Please use `constexpr std::string` or `char[]`

https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

File: sentinel-core/transport/command/command_request.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+ private:
+  std::map<std::string, std::string> metadata_;
```
Comment (zyfjeff) [inline sentinel-core/transport/command/command_request.h]: unordered_map?

File: sentinel-core/transport/command/command_request.cc [hunk 8]
```diff
@@ -0,0 +1,70 @@
+#include "sentinel-core/transport/command/command_request.h"
+
+namespace Sentinel {
+namespace Transport {
+
+const std::string CommandRequest::kRequestTarget = "command-target";
+
+const std::string& CommandRequest::GetBody() const { return body_; }
```
Comment (zyfjeff) [inline sentinel-core/transport/command/command_request.cc]: Please use body、set_body

https://google.github.io/styleguide/cppguide.html#Function_Names

File: sentinel-core/transport/command/command_response.h [hunk 9]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <string>
+
+class CommandResponse {
+ private:
+  CommandResponse(bool success, const std::string& result) : result_(result) {
```
Comment (zyfjeff) [inline sentinel-core/transport/command/command_response.h]: sccess_(success)

File: sentinel-core/transport/command/http_command_center.cc [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+namespace Sentinel {
+namespace Transport {
+
+bool HttpCommandCenter::Start() {
+  if (http_server_) {
+    // log warn
+    return false;
+  }
+
+  http_server_.reset(new HttpServer(std::bind(&HttpCommandCenter::OnHttpRequest,
```
Comment (zyfjeff) [inline sentinel-core/transport/command/http_command_center.cc]: use `std::make_unique< HttpServer>`

File: sentinel-core/log/metric_reader.cc [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    // currentSecond should >= beginSecond, otherwise a wrong metric file must
+    // occur
+
+    if (cur_second < begin_second) {
+      return false;
+    }
+    if (cur_second <= end_second) {
+      // read all
+      if (identity.size() == 0) {
+        metric_vec.push_back(node);
```
Comment (zyfjeff) [inline sentinel-core/log/metric_reader.cc]: use `emplac_back(std::move(node))` to avoid copy

File: sentinel-core/log/metric_reader.cc [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+      } else if (node.GetResource() == identity) {
+        metric_vec.push_back(node);
```
Comment (zyfjeff) [inline sentinel-core/log/metric_reader.cc]: ditto

File: sentinel-core/log/metric_reader.cc [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    } else {
+      return false;
+    }
+
+    if (metric_vec.size() >= kMaxLinesReturn) {
+      return false;
+    }
+  }
+
+  in_file.close();
```
Comment (zyfjeff) [inline sentinel-core/log/metric_reader.cc]: Don't need to call the close method

File: sentinel-core/log/metric_reader.cc [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+#include <fstream>
+#include <iostream>
+#include <istream>
+
+namespace Sentinel {
+namespace Log {
+
+std::vector<Metric::MetricNode> MetricReader::ReadMetrics(
+    std::vector<std::string> file_names, int pos, int64_t offset,
```
Comment (zyfjeff) [inline sentinel-core/log/metric_reader.cc]: `const std::vector<std::string>& file_names`

File: sentinel-core/log/metric_reader.cc [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+    }
+    last_second = cur_second;
+  }
+
+  in_file.close();
+}
+
+std::vector<Metric::MetricNode> MetricReader::ReadMetricsByEndTime(
+    std::vector<std::string> file_names, int pos, int64_t offset,
+    int64_t begin_time_ms, int64_t end_time_ms, std::string identity) {
```
Comment (zyfjeff) [inline sentinel-core/log/metric_reader.cc]: `const std::string& identity`

File: sentinel-core/log/metric_reader.cc [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+                                       begin_time_ms, end_time_ms, identity)) {
+  }
+
+  return metric_vec;
+}
+
+bool MetricReader::ReadMetricsInOneFileByEndTime(
+    std::vector<Metric::MetricNode> &metric_vec, std::string file_name,
+    int64_t offset, int64_t begin_time_ms, int64_t end_time_ms,
+    std::string identity) {
```
Comment (zyfjeff) [inline sentinel-core/log/metric_reader.cc]: ditto

File: sentinel-core/log/metric_searcher.cc [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+  int64_t sec;
+  in >> sec;
+  in.close();
+
+  // timestamp(second) in the specific position == that we cached
+  return sec == last_pos_.second;
+}
+
+int64_t MetricSearcher::FindOffset(int64_t begin_time_ms,
+                                   std::string metric_file_name,
```
Comment (zyfjeff) [inline sentinel-core/log/metric_searcher.cc]: ditto

File: sentinel-core/transport/command/command_handler.h [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Sentinel {
+namespace Transport {
+
+class CommandHandler {
+ public:
+  CommandHandler(const std::string& name) : command_name_(name) {}
+  virtual ~CommandHandler() = default;
+  virtual CommandResponse Handle(const CommandRequest& request) = 0;
+
```
Comment (sczyh30) [inline sentinel-core/transport/command/command_handler.h]: Can we add a `command_name() const` method to get the command name here?

File: sentinel-core/transport/command/http_command_center.cc [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+  return true;
+}
+
+void HttpCommandCenter::OnHttpRequest(struct evhttp_request* http_req) {
+  auto request = HttpCommandUtils::ParseHttpRequest(http_req);
+  auto target = request.GetMetadata(CommandRequest::kRequestTarget);
+
+  auto it = handler_map_.find(target);
+  if (it == handler_map_.end()) {
+    // log warn: can not find handler
```
Comment (sczyh30) [inline sentinel-core/transport/command/http_command_center.cc]: If the handler is absent, maybe we should also return a 400 response to indicate the error like this:

```
Unknown command `abc`
```

File: sentinel-core/log/metric_writer.cc [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+      DoWrite(time, nodes);
+    }
+    last_second_ = second;
+  }
+}
+
+void MetricWriter::DoWrite(int64_t time,
+                           std::vector<Metric::MetricNode> &nodes) {
+  for (auto &node : nodes) {
+    metric_out_ << node.ToFatString();
```
Comment (sczyh30) [inline sentinel-core/log/metric_writer.cc]: We may also need to write `\n` per line.

```suggestion
    metric_out_ << node->ToFatString() << "\n";
```

File: sentinel-core/log/metric_searcher.cc [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+  in.close();
+
+  if (offset != -1) {
+    last_pos_.metric_file_name = metric_file_name;
+    last_pos_.index_file_name = idx_file_name;
+  }
+
+  return offset;
+}
+
```
Comment (sczyh30) [inline sentinel-core/log/metric_searcher.cc]: Where is the original `findByTimeAndResource` method?

File: sentinel-core/utils/file_utils.cc [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    files.push_back(ent->d_name);
+  }
+  closedir(dir);
+
+  return files;
+}
+
+std::string FileUtils::GetAbsolutePath(const std::string& path) {
+  char abs_path[40960] = {0};
+  realpath(path.c_str(), abs_path);
```
Comment (zyfjeff) [inline sentinel-core/utils/file_utils.cc]: check return value?

File: sentinel-core/utils/file_utils.cc [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+  while ((ent = readdir(dir)) != nullptr) {
+    files.push_back(ent->d_name);
+  }
+  closedir(dir);
+
+  return files;
+}
+
+std::string FileUtils::GetAbsolutePath(const std::string& path) {
+  char abs_path[40960] = {0};
```
Comment (zyfjeff) [inline sentinel-core/utils/file_utils.cc]: too large stack size, use 8192 replace 40960?

File: sentinel-core/utils/file_utils.cc [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+  std::vector<std::string> files;
+
+  auto* dir = opendir(path.c_str());
+  if (dir == nullptr) {
+    return files;
+  }
+
+  struct dirent* ent;
+  while ((ent = readdir(dir)) != nullptr) {
+    files.push_back(ent->d_name);
```
Comment (zyfjeff) [inline sentinel-core/utils/file_utils.cc]: use emplace_back

File: sentinel-core/utils/file_utils.cc [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+
+  bool dir_exists = (nullptr != dir);
+  if (dir_exists) {
+    ::closedir(dir);
+  }
+
+  return dir_exists;
+}
+
+int FileUtils::CreateDir(const std::string& path) {
```
Comment (zyfjeff) [inline sentinel-core/utils/file_utils.cc]: s/int/bool?

File: sentinel-core/utils/BUILD [hunk 26]
```diff
@@ -9,6 +9,8 @@ cc_library(
     "utils.cc",
     "time_utils.h",
     "time_utils.cc",
+    "file_utils.h",
```
Comment (zyfjeff) [inline sentinel-core/utils/BUILD]: In a separate cc_library

File: sentinel-core/utils/file_utils.cc [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+namespace Sentinel {
+namespace Utils {
+
+bool FileUtils::FileExists(const std::string& path) {
+  std::ifstream input_file(path);
+  return input_file.is_open();
+}
+
+bool FileUtils::DirExists(const std::string& path) {
+  DIR* const dir = ::opendir(path.c_str());
```
Comment (zyfjeff) [inline sentinel-core/utils/file_utils.cc]: If `opendir` returns `nullptr`, it does not mean that the directory does not exist, there may be a permission problem, the error code needs to be verified is ENOENT

File: sentinel-core/utils/file_utils.cc [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+
+#include <fstream>
+#include <iostream>
+
+namespace Sentinel {
+namespace Utils {
+
+bool FileUtils::FileExists(const std::string& path) {
+  std::ifstream input_file(path);
+  return input_file.is_open();
```
Comment (zyfjeff) [inline sentinel-core/utils/file_utils.cc]: `is_open` fails and there is no guarantee that the file does not exist?

File: sentinel-core/transport/common/event_loop_thread.cc [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+  }
+}
+
+void EventLoopThread::RunTask(Functor func) {
+  {
+    std::lock_guard<std::mutex> lock(task_mutex_);
+    pending_tasks_.emplace_back(func);
+  }
+
+  Wakeup();
```
Comment (zyfjeff) [inline sentinel-core/transport/common/event_loop_thread.cc]: If `DoPendingTasks` is being executed at this time, you should not execute `Wakeup`

File: sentinel-core/transport/command/http_server.h [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+ private:
+  EventLoopThread event_loop_thread_;
+
+  struct evhttp *http_ = nullptr;
+  http_request_callback_t request_callback_;
+
+  static void HttpGenCallback(struct evhttp_request *req, void *arg);
+
+ private:
+  static const int kDefaultPort = 8731;
```
Comment (zyfjeff) [inline sentinel-core/transport/command/http_server.h]: Can I modify this default port?

File: sentinel-core/log/log_base.h [hunk 31]
```diff
@@ -1,5 +1,33 @@
...
+  static void Initialize();
+  static bool IsLogNameUsePid() { return log_name_use_pid_; }
+
+  static std::string GetLogBaseDir() { return log_base_dir_; }
+
+ private:
+  static std::string AddSeparator(std::string dir);
+
+ public:
+  static const std::string kEnvLogDir;
```
Comment (zyfjeff) [inline sentinel-core/log/log_base.h]: use `static const char kEnvLogDir[];`  `std::string` is a non-POD type, static initialization will have out of order problems

https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

File: sentinel-core/log/metric_reader.cc [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+                                       begin_time_ms, end_time_ms, identity)) {
+  }
+
+  return metric_vec;
+}
+
+bool MetricReader::ReadMetricsInOneFileByEndTime(
+    std::vector<Stat::MetricItem> &metric_vec, const std::string &file_name,
+    int64_t offset, int64_t begin_time_ms, int64_t end_time_ms,
+    std::string identity) {
```
Comment (zyfjeff) [inline sentinel-core/log/metric_reader.cc]: `const std::string&`

File: sentinel-core/transport/common/event_loop_thread.cc (hunk 29 above)
Comment (interxd) [reply sentinel-core/transport/common/event_loop_thread.cc]: we also should execute  Wakeup

**Comment by sczyh30** [comment]
@zyfjeff Could you please help to fix the CI build failure?

**Comment by zyfjeff** [comment]
@sczyh30 Ok, I will repair the CI problem as soon as possible

**Pull request merged**